src/zen/websocket_server.cpp

```diff
@@ -0,0 +1,1360 @@
+    }
+
+    void send_mempool_update(int size, std::list<uint256>& listHashes)
+    {
+        sendMempoolEvent(size, listHashes, WsEvent::UPDATE_MEMPOOL);
+    }
+
+    void send_peer_update(std::list<std::string>& peers)
+    {
+        sendPeerEvent(peers, WsEvent::UPDATE_PEER);
+    }
+
+    void shutdown()
+    {
+        try
+        {
+            exit_rwhandler_thread_flag = true;
+            if (this->localWs)
+            {
+                LogPrint("ws", "%s():%d - closing socket\n", __func__, __LINE__);
+                this->localWs->next_layer().close();
+            }
+        }
+        catch (std::exception const& e)
+        {
+            LogPrint("ws", "%s():%d - error: %s\n", __func__, __LINE__, e.what());
+        }
+    }
+};
+
+
+static int getblock(const CBlockIndex *pindex, std::string& strHex)
+{
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    {
+        LOCK(cs_main);
+        CBlock block;
+        if (!ReadBlockFromDisk(block, pindex)) {
+            LogPrint("ws", "%s():%d - error: could not read block from disk\n", __func__, __LINE__);
+            return WsHandler::READ_ERROR;
+        }
+        ss << block;
+        strHex = HexStr(ss.begin(), ss.end());
+    }
+    return WsHandler::OK;
+}
+
+
+
+
+static void ws_updatetip(const CBlockIndex *pindex)
+{
+    std::string strHex;
+    int ret = getblock(pindex, strHex);
+    if (ret != WsHandler::OK)
+    {
+        // should not happen
+        LogPrint("ws", "%s():%d - ERROR: can not update tip\n", __func__, __LINE__);
+        return;
+    }
+    {
+        std::unique_lock<std::mutex> lck(wsmtx);
+        if (listWsHandler.size() )
+        {
+            LogPrint("ws", "%s():%d - update tip loop on ws clients\n", __func__, __LINE__);
+            auto it = listWsHandler.begin();
+            while (it != listWsHandler.end())
+            {
+                LogPrint("ws", "%s():%d - call wshandler_send_tip_update to connection[%u]\n", __func__, __LINE__, (*it)->t_id);
+                (*it)->send_tip_update(pindex->nHeight, pindex->GetBlockHash().GetHex(), strHex, pindex);
+                ++it;
+            }
+        }
+        else
+        {
+            LogPrint("ws", "%s():%d - there are no connected ws clients\n", __func__, __LINE__);
+        }
+    }
+}
+
+static void ws_updatemempool()
+{
+    LOCK(mempool.cs);
+
+    std::list<uint256> hashes;
+    BOOST_FOREACH(const PAIRTYPE(uint256, CTxMemPoolEntry)& entry, mempool.mapTx)
+    {
+       hashes.push_back(entry.first);
+    }
+    {
+		std::unique_lock<std::mutex> lck(wsmtx);
+		if (listWsHandler.size() )
+		{
+			LogPrint("ws", "%s():%d - update mempool loop on ws clients\n", __func__, __LINE__);
+			auto it = listWsHandler.begin();
+			while (it != listWsHandler.end())
+			{
+				LogPrint("ws", "%s():%d - call wshandler_send_mempool_update to connection[%u]\n", __func__, __LINE__, (*it)->t_id);
+				(*it)->send_mempool_update(hashes.size(), hashes);
```

Comment:
It is no use passing also the first parameters: it can be got from the list itself.